File: Solutions/Repeated_Substring_Pattern/Repeated_Substring_Pattern.cpp
//Fastest
class Solution {
public:
    bool repeatedSubstringPattern(string s) {
        int n = s.size();
        if (n < 2) return false;
        if (n == 2) return s[0] == s[1];    
            
        string_view v(s);
        
        vector<int> divisors;
        
        int sqrtN = sqrt(n);
        
        for (int i = 2; i <= sqrtN; ++i) {
            if (n % i == 0) {
                divisors.push_back(i);
                divisors.push_back(n / i);    
            }
        }
        
        divisors.push_back(1);
        
                for (int l : divisors) {
    
                    string_view tmp = v.substr(0, l);
                    string_view first = v.substr(0, l);
                    
                    int firstHash = hash<string_view>{}(tmp);
                    int currHash = firstHash;
                    int start = l;
                    while (start != n && currHash == firstHash) {
                        tmp = v.substr(start, l);
                        currHash = hash<string_view>{}(tmp);
                        start += l;    
                    }
                    if (start == n && currHash == firstHash) {
                        //cout << l << endl;
                        return true;    
                    }    
                }     
               
         
        return false;
    }
};

static auto x = [](){
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);
    std::cout.tie(nullptr);
    return nullptr;
}();

File: Solutions/Repeated_Substring_Pattern/Repeated_Substring_Pattern.cs
public class Solution {
    public bool RepeatedSubstringPattern(string s) {
        
        bool sections = false;
        int size = 0;
        int start =0;
        
        for(int i=2; i<= s.Length; i++)
        {
            if (s.Length % i != 0) { continue; }
            
            sections = true;
            start=0;
            size = s.Length / i;
            
            for(int j=1; j< i && sections; j++)
            {
                sections = sections && CheckSection(s, start, j * size, (j+1) * size); 
                start = j * size;
            }
            
            if (sections) { return true; }
        }
        
        return false;
    }
    
    private bool CheckSection(string s, int start, int mid, int limit)
    {
        for(int i=start, j=mid; j < limit; i++, j++)
        {
            if (s[i] != s[j]) { return false; }
        }
        return true;
    }
}

//Fastest
public class Solution {
    public bool RepeatedSubstringPattern(string s) {
        int n = s.Length;
        
        int[] dp = new int[s.Length];
        
        int j=0;
        
        for(int i=1;i<n;i++)
        {
            j=dp[i-1];
            
            while(j>0&&s[i]!=s[j])
                j=dp[j-1];
            if(s[i]==s[j])
                 j++;
            dp[i]=j;
                
        }
        int len=dp[n-1];
        return len>0&& n%(n-len)==0;
    }
}